apps/transport/lib/validators/enroute_chouette_valid_client.ex

```
@@ -48,28 +48,35 @@ defmodule Transport.EnRouteChouetteValidClient do
 
     case response |> Map.fetch!("user_status") do
       "pending" ->
-        case {get_datetime(response, "created_at"), get_datetime(response, "updated_at")} do
-          {{:ok, created_at, _}, {:ok, updated_at, _}} ->
-            {:pending, DateTime.diff(updated_at, created_at)}
-
-          _ ->
-            :unexpected_datetime_format
+        case get_elapsed(response) do
+          nil -> :unexpected_datetime_format
+          elapsed -> {:pending, elapsed}
         end
 
       "successful" ->
-        {:successful, url}
+        {:successful, url, get_elapsed(response)}
 
       "warning" ->
-        :warning
+        {:warning, get_elapsed(response)}
 
       "failed" ->
-        :failed
+        {:failed, get_elapsed(response)}
 
       _ ->
         :unexpected_validation_status
     end
   end
 
```

Pour ci-dessous : j'ai tendance à préférer des appels en bang "!" qui lèvent une erreur, pour réduire le code, et du coup réduire les potentialités de bug dans le code qui analyse les retours.

Par exemple ici j'aurais opté pour quelque chose comme:

```elixir
defp get_elapsed!(response) do
  created_at = get_datetime!(response, "created_at")
  updated_at = get_datetime!(response, "updated_at")
  DateTime.diff(updated_at, created_at)
end
```

Et (voir https://hexdocs.pm/elixir/1.17.2/Date.html#from_iso8601!/2)

```elixir
defp get_datetime!(map, key) do
  map 
  |> Map.fetch!(key)
  |> DateTime.from_iso8601!()
end
```

C'est d'ailleurs en phase avec le fait que `Map.fetch!` va de toute façon lever une exception en cas de clé manquante, et adopte donc plutôt un style "fail fast" aussi.

(Et du coup, gérer l'erreur dans l'appelant de `get_elapsed!`)

Après ça reste aussi une opinion, mais perso je trouve que ça fait au final beaucoup moins de code sur une codebase quand c'est applicable (il faut que l'appelant puisse intercepter l'erreur, par contre !)

---

> il faut que l'appelant puisse intercepter l'erreur, par contre !

Oui, et tu peux avoir des cas où la clef est manquante/invalide sans que ça soit grave, pas forcément la responsabilité du code appelant. En effet si j'étais cohérent j'aurais pas utilisé `Map.fetch!`. En pratique je peux faire confiance au fait que le datetime est bien formatté. Ce cas d'erreur spécifique m'ennuie.

Je dois revoir cette partie de toute façon, j'ai identifié un problème en retestant avec la validation OnDemand.

---

> Et (voir https://hexdocs.pm/elixir/1.17.2/Date.html#from_iso8601!/2)

Ce helper est fourni pour Date mais pas pour DateTime ; je vais l'émuler.

---

J'imagine que c'est lié au fait que DateTime.from_iso8601 renvoie et le datetime et l'offset, et donc qu'il faudrait l'interpréter également (ce que je ne fais pas, je fais l'hypothèse que l'offset est constant).